packages/web-components/fast-foundation/src/number-field/number-field.spec.ts

```diff
@@ -831,6 +831,86 @@ describe("NumberField", () => {
         });
     });
 
+    describe("value validation", () => {
+        it("should allow number entry", async () => {
+            const { element, connect, disconnect } = await setup();
+
+            await connect();
+
+            element.setAttribute("value", "18");
+
+            expect(element.value).to.equal("18");
+        });
+
+        it("should not allow non-number entry", async () => {
+            const { element, connect, disconnect } = await setup();
+
+            await connect();
+
+            element.setAttribute("value", "11a");
+
+            expect(element.value).to.equal("11");
+        });
+
+        it("should allow float number entry", async () => {
+            const { element, connect, disconnect } = await setup();
+
+            await connect();
```

Comment:
I'm noticing that this implementation does not have behavioral parity with the native `<input type="number" />` when it comes to these tests: https://stackblitz.com/edit/web-platform-sre4og?file=index.html Some of the native behaviors seem a little odd to me (such as "+3" attribute resulting in empty string whereas typing it as a user results in the value '3' ) but others seem intuitive to me ("-", ".", and "+" all emit empty string values); What was the impetus for emitting those three as stateful values?

Reply:
It's not strictly necessary. I was adding for the negative scenario and added positive for completeness.

As for entering "-" and ".", it might make sense to parseFloat the result to the proxy value. This will allow the user to start a value with "-" or "." but the stored proxy will be empty string in the case that the form gets submitted.

Reply:
I don't think that approach will work. The proxy is only used for form-submission in cases where `ElementInternals` is not supported, so this will lead to a mis-match in behavior between different browsers. It would also make `FASTNumberField.value` inconsistent with the value submitted with the form for those browsers, which I think would be unintuitive for authors.

Can we store the typed / attribute / displayed value in a different property and set the stateful value as the result of parseFloat of that value (or empty string if the result is `NaN`)?

Reply:
I'm not suggesting we don't show what was typed to the user. I'm wondering if we can divorce what is *displayed* from the `FASTNumberField.value`. For example:

1. The user types "." so that "." is displayed
2. -> FASTNumberField.value is "" because "." cannot be successfully parsed to a number
3. The user types "3" so that ".3" is displayed
4. -> FASTNumberField.value is now ".3" because ".3" can successfully be parsed to a number
5. The user moves the cursor and deletes the "." so that "3" is displayed
6. -> FASTNumberField.value is now "3" because "3" can successfully be parsed to a number

The above seems to me to be how the native control functions.

Reply:
FWIW, I agree with @nicholasrice on how the value updates according to a user's input. Note that this would also allow features such as numeric formatting, where the displayed value might be "1M", and the FASTNumberField.value is 1000000 (and the user could have entered either "1M" or 1000000 for the same result).

Reply:
I've updated it to use a displayText for the input field. This allows for typing out negative values, decimal values and even exponentials.